uefi-test-runner/src/main.rs

```diff
@@ -10,13 +10,18 @@ extern crate log;
 #[macro_use]
 extern crate alloc;
 
+extern crate x86_64;
+
 use uefi::prelude::*;
 
 mod boot;
 mod proto;
 
 #[no_mangle]
 pub extern "C" fn uefi_start(_handle: uefi::Handle, st: &'static SystemTable) -> Status {
+    // Set up a hook to shut down QEMU when a panic occurs
+    unsafe { uefi_services::set_panic_shutdown_hook(&qemu_panic_shutdown) }
```

Comment:
Bikeshed: could also be an optional parameter to uefi_services::init. Any preference?

Reply:
One nice part to this design is that it makes sure that the user will think about the hook. OTOH, one argument in favor of keeping a separate setter for the hook is that it allows the hook to be changed later on (e.g. after setting up an ACPI driver).

Reply:
The idea is that in a real OS, you wouldn't use `uefi_services` and instead handle everything by yourself.

- Since the `log` crate only allows setting the logger once, once you get a real kernel ringbuffer for the system log, you won't be able to redirect log messages to it.
- Same goes for the global memory allocator, cannot be changed afterwards
- It makes sense to put this stuff in `uefi_services::init`, and keep it one-time-settable only, like its kin

The idea of `uefi_services` is to help users if: 
- their OS doesn't yet have a ringbuffer / memory allocator
- their OS has these features, and they're handled from the early boot by the kernel
- their OS has these features, but the UEFI loader is a separate binary from the kernel

i.e. `uefi_services` is for writing something like the `EFISTUB` part of the Linux kernel, for the early boot; but the kernel itself is a separate binary, glued together with the stub. I will update the docs to make these issues clear.